Disable lazy discover in xiaomi_miio

mrwogu opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds a fix to Xiaomi Miio integration which disable `lazy_discover` mode, which is supported by python-miio.
**This changes the default behavior as it defaulted previously to enabled. If this change breaks user installations during the beta, we should revert it back to enabled.**
Disabling this mode resolves network connection or timeout issues on some Xiaomi devices.

Currently, python-miio supports this for all devices (except RoborockVacuum). For that, we're waiting for python-miio 0.6 release with merged PR https://github.com/rytilahti/python-miio/pull/1606.
For now, I left the TODO in the code.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #59215 #67926
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

mrwogu committed:
Add lazy discover config option to xiaomi_miio (#59215)

rytilahti commented on homeassistant/components/xiaomi_miio/translations/pl.json:
```diff
@@ -58,7 +58,8 @@
                 "data": {
-                    "cloud_subdevices": "U\u017cyj chmury, aby uzyska\u0107 pod\u0142\u0105czone podurz\u0105dzenia"
+                    "cloud_subdevices": "U\u017cyj chmury, aby uzyska\u0107 pod\u0142\u0105czone podurz\u0105dzenia",
+                    "lazy_discover": "U\u017cyj leniwego wykrywania, aby zoptymalizowa\u0107 nawi\u0105zywanie po\u0142\u0105czenia z urz\u0105dzeniem"
```
Only the english version of strings should be commited in, the rest will come automatically from lokalise as they get translated.

rytilahti reviewed (commented):
I'm fine with this, but I'm wondering if we should default to `False` at least for the beta? If we receive bug reports about issues, we can always revert it back to `True` during the beta phase.

My reasoning is that I cannot anymore remember why the lazy discovery was even added to the library, but what it does is, IIRC, is to perform the handshake only 1) if it wasn't ever done before or 2) after an error with communicating the device occurred. If it works fine when it's false, I could remove the setting altogether (or default to False) in the future python-miio release.

I'll do some testing with my few devices using `False`, and report back my findings soonish.

mrwogu committed:
disable lazy_discover for default

mrwogu commented:
I've fixed code with your review comments, and added new commit with change default lazy_discover to `False`.
I'm ok with that change for beta release.
I assumed a non-invasive approach, because not everyone had a problem with the timeout. This is really a non-deterministic error.

Thanks

rytilahti commented:
Great, I modified the description and added second-opinion-wanted to get also someone else to chime in if the approach is viable. Please fix the config flow tests to make the CI pass again.

mrwogu committed:
extend tests to support lazy_discover

rytilahti commented:
Just for the reference, I did some digging to find out that the lazy discovery was added in https://github.com/rytilahti/python-miio/pull/215 and I noted that some other impl. (likely aholstenson's miio lib) performed the handshaking only every 180 or 200 seconds.

I am still perplexed why not having the handshake is causing the issues with the integration given a handshake should be done automatically when encountering an error, but I start to feel that instead of adding a setting we should just default the `lazy_handshake` to `False`.

For 0.6 series of python-miio, I think it the option will go away in favor of "handshake interval" setting that will default to 120s, if the experiences from beta look promising.

starkillerOG commented:
@rytilahti I think sending the handshake every 120 seconds is indeed better than doing it always.
When updating the status of a device like roborock, multiple commands are send sequentially for one call. In those cases otherwise a handshake will happen in between each command, which is a probablly a bit of a waste of IO.

rytilahti commented:
@starkillerOG you are right, I added a todo to https://github.com/rytilahti/python-miio/issues/1114 to replace the current lazy discovery in favor of "time since last handshake" option. If some devices require this for every request, it can be configured to 0.

The beta cut off is on upcoming wednesday, hopefully we can get this PR merged for that.

mrwogu commented:
@rytilahti what I'm supposed to do now? Remove setting and just pass False as argument to lazy_discovery?

rytilahti commented:
@mrwogu yeah, I think it would be the best option for the time being, and other devs agreed on discord that introducing low-level configuration options is not a way to go.

I cannot test this currently against the devices in my possession, but we are going to need more input to verify that it doesn't break the integration against other devices. Do you have other devices besides the air purifier to check that they keep working? 

@starkillerOG would you have some time to test the change against the devices you have?

mrwogu committed:
revert config option

mrwogu commented:
@rytilahti done.

> Do you have other devices besides the air purifier to check that they keep working?

Yes, I have several purifiers, a humidifier, a fan and a Roborock. 
My PR made the problems disappear. Unfortunately, my PR won't fix Roborock, and here it is also a nuisance.
I have to use the python-miio version I built myself.
That's why I'm quietly hoping that maybe it would be possible to release a python-miio hotfix with a corrected constructor.

rytilahti commented:
> Yes, I have several purifiers, a humidifier, a fan and a Roborock.

Okay, great!

> My PR made the problems disappear. Unfortunately, my PR won't fix Roborock, and here it is also a nuisance.
> That's why I'm quietly hoping that maybe it would be possible to release a python-miio hotfix with a corrected constructor.

So, my personal preference–given the current ongoing refactoring of the lib & the integration–would be disabling it also for the vacuum by setting the property just like `dev._protocol.lazy_discover = False`.

I created https://github.com/rytilahti/python-miio/pull/1606 to fix the constructor signature, but I don't have any free time to create a separate maintenance branch for 0.5 series and prepare releases, etc. The 0.6 push is much more important for future-proofing the library, and that's where I prioritize my personal efforts at the moment.

starkillerOG commented:
I can indeed test with my roborock and aqara gateways, but the current code does not touch the vacuums or gateways, so there is now no point in testing for me.

frenck commented:
@rytilahti What are the next steps needed to move this PR forward?

rytilahti commented:
> So, my personal preference–given the current ongoing refactoring of the lib & the integration–would be disabling it also for the vacuum by setting the property just like dev._protocol.lazy_discover = False.

@frenck Basically doing this & getting this merged for broader testing would be my preference. If it causes more trouble than it fixes, we can easily revert/flip the setting.

Maeur1 commented:
Friendly ping on this PR (if its still relevant for https://github.com/home-assistant/core/issues/59215). I have this with one of my fans, it constantly is timing out and would be happy to test this fix out if guided on how to run this locally?

rytilahti reviewed (approved):
I think we should just merge this, and revert if it causes more troubles. In the best case this at least alleviates the issue encountered by so many. Thanks for the PR @mrwogu, let's hope this fixes the issue for the time being! :1st_place_medal:
